File: src/azure_health_beacon/updater.py
from __future__ import annotations

import hashlib
import json
import os
import re
import subprocess
import tempfile
from dataclasses import dataclass
from pathlib import Path
from typing import Any, BinaryIO
from urllib.parse import urlparse
from urllib.request import Request, urlopen

from . import __version__
from .config import app_data_dir

REPOSITORY = "Anders0lesen/azure-extensible-systray-monitor"
LATEST_RELEASE_API = f"https://api.github.com/repos/{REPOSITORY}/releases/latest"
RELEASE_PAGE_PREFIX = f"https://github.com/{REPOSITORY}/releases/tag/"
DOWNLOAD_PREFIX = f"/{REPOSITORY}/releases/download/"
MAX_RELEASE_JSON_BYTES = 1_000_000
MAX_CHECKSUM_BYTES = 4096
MAX_INSTALLER_BYTES = 100_000_000
ALLOWED_FINAL_DOWNLOAD_HOSTS = {
    "github.com",
    "objects.githubusercontent.com",
    "release-assets.githubusercontent.com",
}


@dataclass(frozen=True, slots=True)
class ReleaseInfo:
    version: str
    tag: str
    title: str
    notes: str
    page_url: str
    installer_name: str
    installer_url: str
    installer_digest: str
    checksum_name: str
    checksum_url: str


def parse_version(value: str) -> tuple[int, int, int]:
    match = re.fullmatch(r"v?(0|[1-9]\d*)\.(0|[1-9]\d*)\.(0|[1-9]\d*)", value)
    if not match:
        raise ValueError(f"Unsupported version: {value}")
    return tuple(int(part) for part in match.groups())  # type: ignore[return-value]


def is_newer_version(candidate: str, current: str = __version__) -> bool:
    return parse_version(candidate) > parse_version(current)


def _read_limited(response: BinaryIO, limit: int) -> bytes:
    declared = response.headers.get("Content-Length")
    if declared and int(declared) > limit:
        raise ValueError("The update response exceeds its safety limit")
    data = response.read(limit + 1)
    if len(data) > limit:
        raise ValueError("The update response exceeds its safety limit")
    return data


def _request(url: str, timeout_seconds: int) -> Any:
    request = Request(
        url,
        headers={
            "Accept": "application/vnd.github+json",
            "User-Agent": f"AzureHealthBeacon/{__version__}",
            "X-GitHub-Api-Version": "2022-11-28",
        },
    )
    return urlopen(request, timeout=timeout_seconds)  # noqa: S310 - URLs are pinned and validated.


def _validate_release_asset_url(url: str, tag: str, asset_name: str) -> None:
    parsed = urlparse(url)
    expected_path = f"{DOWNLOAD_PREFIX}{tag}/{asset_name}"
    if (
        parsed.scheme != "https"
        or parsed.hostname != "github.com"
        or parsed.username is not None
        or parsed.password is not None
        or parsed.port not in (None, 443)
        or parsed.path != expected_path
        or parsed.query
        or parsed.fragment
    ):
        raise ValueError("GitHub returned an unexpected update download URL")


def parse_release_payload(payload: dict[str, Any]) -> ReleaseInfo:
    if payload.get("draft") or payload.get("prerelease"):
        raise ValueError("The latest GitHub release is not a stable release")
    tag = str(payload.get("tag_name", ""))
    version = tag.removeprefix("v")
    parse_version(version)
    page_url = str(payload.get("html_url", ""))
    if page_url != f"{RELEASE_PAGE_PREFIX}{tag}":
        raise ValueError("GitHub returned an unexpected release page")

    installer_name = f"AzureHealthBeacon-Setup-{tag}.exe"
    checksum_name = f"{installer_name}.sha256"
    assets = payload.get("assets")
    if not isinstance(assets, list):
        raise TypeError("The GitHub release contains no downloadable assets")
    by_name = {
        str(asset.get("name", "")): asset for asset in assets if isinstance(asset, dict)
    }
    if installer_name not in by_name or checksum_name not in by_name:
        raise ValueError("The release is missing its installer or checksum")
    installer_url = str(by_name[installer_name].get("browser_download_url", ""))
    checksum_url = str(by_name[checksum_name].get("browser_download_url", ""))
    installer_digest = str(by_name[installer_name].get("digest", ""))
    if not re.fullmatch(r"sha256:[a-fA-F0-9]{64}", installer_digest):
        raise ValueError("The release installer has no GitHub SHA-256 digest")
    _validate_release_asset_url(installer_url, tag, installer_name)
    _validate_release_asset_url(checksum_url, tag, checksum_name)
    return ReleaseInfo(
        version=version,
        tag=tag,
        title=str(payload.get("name") or tag)[:200],
        notes=str(payload.get("body") or "")[:20_000],
        page_url=page_url,
        installer_name=installer_name,
        installer_url=installer_url,
        installer_digest=installer_digest.removeprefix("sha256:").casefold(),
        checksum_name=checksum_name,
        checksum_url=checksum_url,
    )


def fetch_latest_release(timeout_seconds: int = 15) -> ReleaseInfo:
    with _request(LATEST_RELEASE_API, timeout_seconds) as response:
        final = urlparse(response.geturl())
        if final.scheme != "https" or final.hostname != "api.github.com":
            raise ValueError("The update service redirected outside GitHub")
        raw = _read_limited(response, MAX_RELEASE_JSON_BYTES)
    payload = json.loads(raw.decode("utf-8"))
    if not isinstance(payload, dict):
        raise TypeError("GitHub returned an invalid release document")
    return parse_release_payload(payload)


def _expected_checksum(text: str, installer_name: str) -> str:
    line = text.strip()
    match = re.fullmatch(r"([a-fA-F0-9]{64})\s+\*?([^\r\n]+)", line)
    if not match or match.group(2) != installer_name:
        raise ValueError("The release checksum file is invalid")
    return match.group(1).casefold()


def _validate_final_download(response: Any) -> None:
    parsed = urlparse(response.geturl())
    if (
        parsed.scheme != "https"
        or parsed.hostname not in ALLOWED_FINAL_DOWNLOAD_HOSTS
        or parsed.username is not None
        or parsed.password is not None
        or parsed.port not in (None, 443)
    ):
        raise ValueError("The update download redirected outside GitHub")


def download_verified_installer(
    release: ReleaseInfo,
    target_directory: Path | None = None,
    timeout_seconds: int = 60,
) -> Path:
    target = target_directory or app_data_dir() / "updates"
    target.mkdir(parents=True, exist_ok=True)

    with _request(release.checksum_url, timeout_seconds) as response:
        _validate_final_download(response)
        checksum_text = _read_limited(response, MAX_CHECKSUM_BYTES).decode("ascii")
    expected = _expected_checksum(checksum_text, release.installer_name)
    if expected != release.installer_digest:
        raise ValueError("The published checksums for the installer do not agree")

    handle, temporary_name = tempfile.mkstemp(
        prefix="update-", suffix=".part", dir=target
    )
    digest = hashlib.sha256()
    total = 0
    try:
        with os.fdopen(handle, "wb") as stream:
            with _request(release.installer_url, timeout_seconds) as response:
                _validate_final_download(response)
                declared = response.headers.get("Content-Length")
                if declared and int(declared) > MAX_INSTALLER_BYTES:
                    raise ValueError("The update installer exceeds its safety limit")
                while chunk := response.read(1024 * 1024):
                    total += len(chunk)
                    if total > MAX_INSTALLER_BYTES:
                        raise ValueError(
                            "The update installer exceeds its safety limit"
                        )
                    digest.update(chunk)
                    stream.write(chunk)
        if total == 0 or digest.hexdigest().casefold() != expected:
            raise ValueError("The downloaded installer failed SHA-256 verification")
        destination = target / release.installer_name
        os.replace(temporary_name, destination)
        return destination
    finally:
        if os.path.exists(temporary_name):
            os.unlink(temporary_name)


def launch_installer(installer: Path, *, automatic: bool) -> None:
    # Both an explicitly approved update and opt-in automatic updating are
    # in-place operations. The surrounding UI owns consent and error reporting.
    arguments = [
        str(installer),
        "/VERYSILENT",
        "/SUPPRESSMSGBOXES",
        "/NORESTART",
        "/CLOSEAPPLICATIONS",
        "/RESTARTAPPLICATIONS",
    ]
    subprocess.Popen(arguments, cwd=installer.parent, close_fds=True)  # noqa: S603

File: src-tauri/src/updater.rs
use std::{
    fs,
    io::{Read, Write},
    path::{Path, PathBuf},
    process::Command,
    time::Duration,
};

use reqwest::{
    StatusCode,
    blocking::{Client, Response},
    header::LOCATION,
};
use serde::Serialize;
use serde_json::Value;
use sha2::{Digest, Sha256};
use url::Url;
use uuid::Uuid;

use crate::{config::app_data_dir, security::atomic_replace};

const LATEST_RELEASE_API: &str =
    "https://api.github.com/repos/Anders0lesen/azure-extensible-systray-monitor/releases/latest";
const RELEASE_PAGE_PREFIX: &str =
    "https://github.com/Anders0lesen/azure-extensible-systray-monitor/releases/tag/";
const DOWNLOAD_PATH_PREFIX: &str =
    "/Anders0lesen/azure-extensible-systray-monitor/releases/download/";
const MAX_RELEASE_BYTES: usize = 1_000_000;
const MAX_CHECKSUM_BYTES: usize = 4_096;
const MAX_INSTALLER_BYTES: u64 = 100_000_000;

#[derive(Clone, Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct ReleaseInfo {
    pub update_available: bool,
    pub version: String,
    pub tag: String,
    pub title: String,
    pub notes: String,
    pub page_url: String,
    #[serde(skip_serializing)]
    installer_name: String,
    #[serde(skip_serializing)]
    installer_url: Url,
    #[serde(skip_serializing)]
    checksum_url: Url,
    #[serde(skip_serializing)]
    installer_digest: String,
}

pub fn check(current_version: &str, timeout_seconds: u64) -> Result<ReleaseInfo, String> {
    let client = client(timeout_seconds)?;
    let response = client
        .get(LATEST_RELEASE_API)
        .header("Accept", "application/vnd.github+json")
        .header("X-GitHub-Api-Version", "2022-11-28")
        .header("User-Agent", format!("AzureHealthBeacon/{current_version}"))
        .send()
        .map_err(|_| "GitHub could not be reached securely")?;
    if !response.status().is_success() {
        return Err(format!("GitHub returned HTTP {}", response.status()));
    }
    let bytes = limited_bytes(response, MAX_RELEASE_BYTES)?;
    let payload: Value =
        serde_json::from_slice(&bytes).map_err(|_| "GitHub returned invalid release metadata")?;
    parse_release(&payload, current_version)
}

pub fn download(release: &ReleaseInfo, timeout_seconds: u64) -> Result<PathBuf, String> {
    validate_release_url(
        &release.installer_url,
        &release.tag,
        &release.installer_name,
    )?;
    validate_release_url(
        &release.checksum_url,
        &release.tag,
        &format!("{}.sha256", release.installer_name),
    )?;
    let checksum = String::from_utf8(download_limited(
        release.checksum_url.clone(),
        timeout_seconds,
        MAX_CHECKSUM_BYTES as u64,
    )?)
    .map_err(|_| "The release checksum is not plain text")?;
    let expected = parse_checksum(&checksum, &release.installer_name)?;
    if expected != release.installer_digest {
        return Err("GitHub's installer digest and checksum asset do not agree".into());
    }

    let directory = app_data_dir().join("updates");
    fs::create_dir_all(&directory).map_err(|_| "The update directory could not be created")?;
    let temporary = directory.join(format!("update-{}.part", Uuid::new_v4()));
    let outcome = (|| {
        let client = client(timeout_seconds)?;
        let mut response = get_following_safe_redirects(&client, release.installer_url.clone())?;
        if response
            .content_length()
            .is_some_and(|value| value > MAX_INSTALLER_BYTES)
        {
            return Err("The update installer exceeds the 100 MB safety limit".into());
        }
        let mut file = fs::File::create(&temporary)
            .map_err(|_| "The temporary update installer could not be created")?;
        let mut hasher = Sha256::new();
        let mut total = 0_u64;
        let mut buffer = [0_u8; 64 * 1024];
        loop {
            let count = response
                .read(&mut buffer)
                .map_err(|_| "The update download stopped unexpectedly")?;
            if count == 0 {
                break;
            }
            total = total.saturating_add(count as u64);
            if total > MAX_INSTALLER_BYTES {
                return Err("The update installer exceeds the 100 MB safety limit".into());
            }
            hasher.update(&buffer[..count]);
            file.write_all(&buffer[..count])
                .map_err(|_| "The update installer could not be written")?;
        }
        file.sync_all()
            .map_err(|_| "The update installer could not be flushed")?;
        let digest = hasher.finalize();
        if total == 0 || hex_digest(&digest) != expected {
            return Err("The downloaded installer failed SHA-256 verification".into());
        }
        let destination = directory.join(&release.installer_name);
        atomic_replace(&temporary, &destination)?;
        Ok(destination)
    })();
    let _ = fs::remove_file(&temporary);
    outcome
}

pub fn launch(installer: &Path) -> Result<(), String> {
    let parent = installer
        .parent()
        .ok_or("The update installer path is incomplete")?;
    Command::new(installer)
        .current_dir(parent)
        .args([
            "/VERYSILENT",
            "/SUPPRESSMSGBOXES",
            "/NORESTART",
            "/CLOSEAPPLICATIONS",
            "/RESTARTAPPLICATIONS",
        ])
        .spawn()
        .map_err(|_| "The verified update installer could not be started")?;
    Ok(())
}

fn parse_release(payload: &Value, current: &str) -> Result<ReleaseInfo, String> {
    if payload
        .get("draft")
        .and_then(Value::as_bool)
        .unwrap_or(true)
        || payload
            .get("prerelease")
            .and_then(Value::as_bool)
            .unwrap_or(true)
    {
        return Err("The latest GitHub release is not stable".into());
    }
    let tag = payload
        .get("tag_name")
        .and_then(Value::as_str)
        .ok_or("GitHub returned no release tag")?;
    let version = tag.strip_prefix('v').unwrap_or(tag);
    let current_parts = version_parts(current)?;
    let candidate_parts = version_parts(version)?;
    let page_url = payload
        .get("html_url")
        .and_then(Value::as_str)
        .unwrap_or_default();
    if page_url != format!("{RELEASE_PAGE_PREFIX}{tag}") {
        return Err("GitHub returned an unexpected release page".into());
    }
    let installer_name = format!("AzureHealthBeacon-Setup-{tag}.exe");
    let checksum_name = format!("{installer_name}.sha256");
    let assets = payload
        .get("assets")
        .and_then(Value::as_array)
        .ok_or("The release has no assets")?;
    let asset = |name: &str| {
        assets
            .iter()
            .find(|item| item.get("name").and_then(Value::as_str) == Some(name))
    };
    let installer = asset(&installer_name).ok_or("The release is missing its Windows installer")?;
    let checksum = asset(&checksum_name).ok_or("The release is missing its checksum")?;
    let digest = installer
        .get("digest")
        .and_then(Value::as_str)
        .unwrap_or_default()
        .strip_prefix("sha256:")
        .unwrap_or_default()
        .to_ascii_lowercase();
    if digest.len() != 64 || !digest.bytes().all(|byte| byte.is_ascii_hexdigit()) {
        return Err("The GitHub installer has no valid SHA-256 digest".into());
    }
    let installer_url = Url::parse(
        installer
            .get("browser_download_url")
            .and_then(Value::as_str)
            .unwrap_or_default(),
    )
    .map_err(|_| "GitHub returned an invalid installer URL")?;
    let checksum_url = Url::parse(
        checksum
            .get("browser_download_url")
            .and_then(Value::as_str)
            .unwrap_or_default(),
    )
    .map_err(|_| "GitHub returned an invalid checksum URL")?;
    validate_release_url(&installer_url, tag, &installer_name)?;
    validate_release_url(&checksum_url, tag, &checksum_name)?;
    Ok(ReleaseInfo {
        update_available: candidate_parts > current_parts,
        version: version.to_owned(),
        tag: tag.to_owned(),
        title: payload
            .get("name")
            .and_then(Value::as_str)
            .unwrap_or(tag)
            .chars()
            .take(200)
            .collect(),
        notes: payload
            .get("body")
            .and_then(Value::as_str)
            .unwrap_or_default()
            .chars()
            .take(20_000)
            .collect(),
        page_url: page_url.to_owned(),
        installer_name,
        installer_url,
        checksum_url,
        installer_digest: digest,
    })
}

fn validate_release_url(url: &Url, tag: &str, name: &str) -> Result<(), String> {
    let expected = format!("{DOWNLOAD_PATH_PREFIX}{tag}/{name}");
    if url.scheme() != "https"
        || url.host_str() != Some("github.com")
        || url.path() != expected
        || url.query().is_some()
        || url.fragment().is_some()
        || !url.username().is_empty()
        || url.password().is_some()
    {
        Err("GitHub returned an unexpected update asset URL".into())
    } else {
        Ok(())
    }
}

fn get_following_safe_redirects(client: &Client, mut url: Url) -> Result<Response, String> {
    for _ in 0..6 {
        let response = client
            .get(url.clone())
            .header("User-Agent", "AzureHealthBeacon/0.8")
            .send()
            .map_err(|_| "The update download could not be reached securely")?;
        if response.status().is_redirection() {
            let location = response
                .headers()
                .get(LOCATION)
                .and_then(|value| value.to_str().ok())
                .ok_or("The update redirect was incomplete")?;
            url = url
                .join(location)
                .map_err(|_| "The update redirect URL was invalid")?;
            validate_download_host(&url)?;
            continue;
        }
        if response.status() != StatusCode::OK {
            return Err(format!(
                "The update download returned HTTP {}",
                response.status()
            ));
        }
        validate_download_host(response.url())?;
        return Ok(response);
    }
    Err("The update download used too many redirects".into())
}

fn validate_download_host(url: &Url) -> Result<(), String> {
    const HOSTS: &[&str] = &[
        "github.com",
        "objects.githubusercontent.com",
        "release-assets.githubusercontent.com",
    ];
    if url.scheme() == "https"
        && url.host_str().is_some_and(|host| HOSTS.contains(&host))
        && url.port().is_none_or(|port| port == 443)
        && url.username().is_empty()
        && url.password().is_none()
    {
        Ok(())
    } else {
        Err("The update download redirected outside approved GitHub hosts".into())
    }
}

fn download_limited(url: Url, timeout: u64, limit: u64) -> Result<Vec<u8>, String> {
    let client = client(timeout)?;
    let response = get_following_safe_redirects(&client, url)?;
    limited_bytes(response, limit as usize)
}

fn limited_bytes(mut response: Response, limit: usize) -> Result<Vec<u8>, String> {
    if response
        .content_length()
        .is_some_and(|value| value > limit as u64)
    {
        return Err("The update response exceeds its safety limit".into());
    }
    let mut bytes = Vec::new();
    response
        .take(limit as u64 + 1)
        .read_to_end(&mut bytes)
        .map_err(|_| "The update response could not be read")?;
    if bytes.len() > limit {
        Err("The update response exceeds its safety limit".into())
    } else {
        Ok(bytes)
    }
}

fn parse_checksum(text: &str, installer_name: &str) -> Result<String, String> {
    let mut parts = text.trim().split_whitespace();
    let digest = parts.next().unwrap_or_default().to_ascii_lowercase();
    let name = parts.next().unwrap_or_default().trim_start_matches('*');
    if parts.next().is_some()
        || digest.len() != 64
        || !digest.bytes().all(|byte| byte.is_ascii_hexdigit())
        || name != installer_name
    {
        Err("The release checksum file is invalid".into())
    } else {
        Ok(digest)
    }
}

fn version_parts(value: &str) -> Result<(u64, u64, u64), String> {
    let values = value
        .strip_prefix('v')
        .unwrap_or(value)
        .split('.')
        .map(str::parse::<u64>)
        .collect::<Result<Vec<_>, _>>()
        .map_err(|_| "The release version is invalid")?;
    if values.len() != 3 {
        return Err("The release version is invalid".into());
    }
    Ok((values[0], values[1], values[2]))
}

fn hex_digest(bytes: &[u8]) -> String {
    bytes.iter().map(|byte| format!("{byte:02x}")).collect()
}

fn client(timeout_seconds: u64) -> Result<Client, String> {
    Client::builder()
        .redirect(reqwest::redirect::Policy::none())
        .timeout(Duration::from_secs(timeout_seconds.clamp(5, 300)))
        .build()
        .map_err(|_| "A secure update client could not be created".into())
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn rejects_lookalike_download_hosts() {
        let url = Url::parse("https://github.com.example.test/file").unwrap();
        assert!(validate_download_host(&url).is_err());
    }

    #[test]
    fn checksum_requires_exact_asset_name() {
        let digest = "a".repeat(64);
        assert!(parse_checksum(&format!("{digest} *expected.exe"), "expected.exe").is_ok());
        assert!(parse_checksum(&format!("{digest} *other.exe"), "expected.exe").is_err());
    }

    #[test]
    fn semantic_versions_are_compared_numerically() {
        assert!(version_parts("0.10.0").unwrap() > version_parts("0.8.9").unwrap());
    }
}
